# EDA

The following documents is the first and general EDA (Examinating Data Analys) of the 17 datasets listed below:
- Alternative Fuel Vehicles US.csv
- Electric and Alternative Fuel Charging Stations.csv
- ElectricCarData_Norm.csv
- ElectricCarData_Clean.csv
- Light Duty Vehicles.csv
- yellow_tripdata_2022-06.parquet
- yellow_tripdata_2023-06.parquet 
- fhv_tripdata_2022-06.parquet
- fhv_tripdata_2023-06.parquet
- green_tripdata_2022-06.parquet
- green_tripdata_2023-06.parquet
- fhvhv_tripdata_2022-06.parquet
- fhvhv_tripdata_2023-06.parquet
- data_reports_monthly.csv
- taxi+_zone_lookup.csv
- taxi_zones.dbf
- Vehicle Fuel Economy Data.csv

## Libraries

In [1]:
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import pyarrow
import math
from sklearn.ensemble import IsolationForest
from dbfread import DBF
import warnings
warnings.filterwarnings("ignore")

## YELLOW TAXIS
### yellow_tripdata_2022-06
Importing the dataset as a DataFrame

Importing the dataset as a DataFrame

In [2]:
df_yel22= pd.read_parquet('Datasets\yellow_tripdata_2022-06.parquet')


In [3]:
type(df_yel22)

pandas.core.frame.DataFrame

### Data Dictionary:

- VendorID: A code indicating the TPEP provider that provided the record. 
  + 1= Creative Mobile Technologies, LLC.
  + 2= VeriFone Inc.
- tpep_pickup_datetime: The date and time when the meter was engaged.
- tpep_dropoff_datetime: The date and time when the meter was disengaged.
- Passenger_count: The number of passengers in the vehicle(This is a driver-entered value).
- Trip_distance: The elapsed trip distance in miles reported by the taximeter.
- PULocationID: TLC Taxi Zone in which the taximeter was engaged
- DOLocationID: TLC Taxi Zone in which the taximeter was disengaged
- RateCodeID: The final rate code in effect at the end of the trip.
   + 1= Standard rate
   + 2=JFK
   + 3=Newark
   + 4=Nassau or Westchester
   + 5=Negotiated fare
   + 6=Group ride
- Store_and_fwd_flag: This flag indicates whether the trip record was held in vehicle memory before sending to the vendor, aka “store and forward,” because the vehicle did not have a connection to the server.
   + Y= store and forward trip
   + N= not a store and forward trip
- Payment_type: A numeric code signifying how the passenger paid for the trip.
   + 1= Credit card
   + 2= Cash
   + 3= No charge
   + 4= Dispute
   + 5= Unknown
   + 6= Voided trip
- Fare_amount: The time-and-distance fare calculated by the meter.
- Extra: Miscellaneous extras and surcharges. Currently, this only includes the $0.50 and $1 rush hour and overnight charges.
- MTA_tax: $0.50 MTA tax that is automatically triggered based on the metered rate in use.
- Improvement_surcharge: $0.30 improvement surcharge assessed trips at the flag drop. The
- improvement surcharge: began being levied in 2015.
- Tip_amount: Tip amount (This field is automatically populated for credit card tips. Cash tips are not included.)
- Tolls_amount: Total amount of all tolls paid in trip.
- Total_amount: The total amount charged to passengers. Does not include cash tips.
- Congestion_Surcharge: Total amount collected in trip for NYS congestion surcharge.
- Airport_fee: $1.25 for pick up only at LaGuardia and John F. Kennedy Airports

### Shape and Dimension

In [4]:
df_yel22.shape

(3558124, 19)

Column Names and Data Types

In [5]:
df_yel22.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3558124 entries, 0 to 3558123
Data columns (total 19 columns):
 #   Column                 Dtype         
---  ------                 -----         
 0   VendorID               int64         
 1   tpep_pickup_datetime   datetime64[us]
 2   tpep_dropoff_datetime  datetime64[us]
 3   passenger_count        float64       
 4   trip_distance          float64       
 5   RatecodeID             float64       
 6   store_and_fwd_flag     object        
 7   PULocationID           int64         
 8   DOLocationID           int64         
 9   payment_type           int64         
 10  fare_amount            float64       
 11  extra                  float64       
 12  mta_tax                float64       
 13  tip_amount             float64       
 14  tolls_amount           float64       
 15  improvement_surcharge  float64       
 16  total_amount           float64       
 17  congestion_surcharge   float64       
 18  airport_fee           

Here we see that there are no string data in any column, we just have numberic, date and time values in this dataset.

### Data Information

Duplicates

In [6]:
df_yel22.duplicated().sum()

0

There are 0 duplicated rows in this dataset

Missing and Null Values

Below is shown that there are five columns with 132448 null values in each.

Incomplete Columns:
- passenger_count
- RatecodeID
- store_and_fwd_flag
- congestion_surcharge
- airport_fee

In [7]:
df_yel22.isnull().sum()

VendorID                      0
tpep_pickup_datetime          0
tpep_dropoff_datetime         0
passenger_count          132448
trip_distance                 0
RatecodeID               132448
store_and_fwd_flag       132448
PULocationID                  0
DOLocationID                  0
payment_type                  0
fare_amount                   0
extra                         0
mta_tax                       0
tip_amount                    0
tolls_amount                  0
improvement_surcharge         0
total_amount                  0
congestion_surcharge     132448
airport_fee              132448
dtype: int64

### yellow_tripdata_2022-06.parquet

Importing the dataset as a DataFrame

In [8]:
df_yel23= pd.read_parquet('Datasets\yellow_tripdata_2023-06.parquet')


In [9]:
type(df_yel23)

pandas.core.frame.DataFrame

### Data Dictionary:

- It is used the same previous data dictonary.

### Shape and Dimension

In [10]:
df_yel23.shape

(3307234, 19)

Column Names and Data Types

In [11]:
df_yel23.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3307234 entries, 0 to 3307233
Data columns (total 19 columns):
 #   Column                 Dtype         
---  ------                 -----         
 0   VendorID               int32         
 1   tpep_pickup_datetime   datetime64[us]
 2   tpep_dropoff_datetime  datetime64[us]
 3   passenger_count        float64       
 4   trip_distance          float64       
 5   RatecodeID             float64       
 6   store_and_fwd_flag     object        
 7   PULocationID           int32         
 8   DOLocationID           int32         
 9   payment_type           int64         
 10  fare_amount            float64       
 11  extra                  float64       
 12  mta_tax                float64       
 13  tip_amount             float64       
 14  tolls_amount           float64       
 15  improvement_surcharge  float64       
 16  total_amount           float64       
 17  congestion_surcharge   float64       
 18  Airport_fee           

Here we see that there are no string data in any column, we just have numberic, date and time values in this dataset.

### Data Information

Duplicates

In [12]:
df_yel23.duplicated().sum()

0

There are 0 duplicated rows in this dataset

Missing and Null Values

Below is shown that there are five columns with 132448 null values in each.

Incomplete Columns:
- passenger_count
- RatecodeID
- store_and_fwd_flag
- congestion_surcharge
- airport_fee

In [13]:
df_yel23.isnull().sum()

VendorID                     0
tpep_pickup_datetime         0
tpep_dropoff_datetime        0
passenger_count          99887
trip_distance                0
RatecodeID               99887
store_and_fwd_flag       99887
PULocationID                 0
DOLocationID                 0
payment_type                 0
fare_amount                  0
extra                        0
mta_tax                      0
tip_amount                   0
tolls_amount                 0
improvement_surcharge        0
total_amount                 0
congestion_surcharge     99887
Airport_fee              99887
dtype: int64

## GREEN TAXIS
### Green_tripdata_2022-06
Importing the dataset as a DataFrame

In [14]:
green22 = pd.read_parquet('Datasets\\green_tripdata_2022-06.parquet')

### Data Dictionary:

- VendorID: A code indicating the TPEP provider that provided the record.
- lpep_pickup_datetime: The date and time when the meter was engaged.
- lpep_dropoff_datetime: The date and time when the meter was disengaged.
- passenger_count: The number of passengers in the vehicle.
- trip_distance: The elapsed trip distance in miles reported by the taximeter.
- RatecodeID: The final rate code in effect at the end of the trip.
- store_and_fwd_flag: This flag indicates whether the trip record was held in vehicle memory before sending to the vendor, aka “store and forward,” because the vehicle did not have a connection to the server.
- PULocationID: TLC Taxi Zone in which the taximeter was engaged
- DOLocationID: TLC Taxi Zone in which the taximeter was disengaged
- payment_type: A numeric code signifying how the passenger paid for the trip.
- fare_amount: The time-and-distance fare calculated by the meter.
- extra: Miscellaneous extras and surcharges.
- MTA_tax: $0.50 MTA tax that is automatically triggered based on the metered rate in use.
- Improvementent_surcharge: $0.30 improvement surcharge assessed trips at the flag drop. The improvement surcharge began being levied in 2015.
- Tip amount: Tip amount – This field is automatically populated for credit card tips. Cash tips are not included.
- Tolls_amount: Total amount of all tolls paid in trip.
- Total_amount: The total amount charged to passengers. Does not include cash tips.
- Congestion_surcharge: Total amount collected in trip for NYS congestion surcharge.
- ehail_fee: Electronic hailing (E-Hail) allows a passenger to use TLC-licensed apps to hail a taxi.
- Trip_type: A code indicating whether the trip was a street-hail or a dispatch
that is automatically assigned based on the metered rate in use but
can be altered by the driver.

In [15]:
green22.head()

VendorID lpep_pickup_datetime lpep_dropoff_datetime store_and_fwd_flag  \
0         2  2022-06-01 00:29:05   2022-06-01 00:33:47                  N   
1         2  2022-06-01 01:01:36   2022-06-01 01:13:58                  N   
2         2  2022-06-01 00:49:20   2022-06-01 00:49:32                  N   
3         2  2022-06-01 00:03:55   2022-06-01 00:15:07                  N   
4         2  2022-06-01 00:35:38   2022-06-01 00:51:39                  N   

   RatecodeID  PULocationID  DOLocationID  passenger_count  trip_distance  \
0         1.0           134           216              2.0           2.03   
1         1.0           134           216              2.0           4.26   
2         5.0            53            53              1.0           0.00   
3         1.0            65            17              1.0           1.78   
4         1.0            41           159              3.0           3.10   

   fare_amount  extra  mta_tax  tip_amount  tolls_amount ehail_fee  \
0          7.5    0.5      0.5        2.64           0.0      None   
1         14.5    0.5      0.5        3.16           0.0      None   
2         15.0    0.0      0.0        4.59           0.0      None   
3          9.5    0.5      0.5        1.00           0.0      None   
4         13.0    0.5      0.5        0.00           0.0      None   

   improvement_surcharge  total_amount  payment_type  trip_type  \
0                    0.3         11.44           1.0        1.0   
1                    0.3         18.96           1.0        1.0   
2                    0.3         19.89           1.0        2.0   
3                    0.3         11.80           1.0        1.0   
4                    0.3         14.30           2.0        1.0   

   congestion_surcharge  
0                   0.0  
1                   0.0  
2                   0.0  
3                   0.0  
4                   0.0

### Shape and Dimension

In [16]:
green22.shape

(73718, 20)

Column Names and Data Types

In [17]:
green22.dtypes

VendorID                          int64
lpep_pickup_datetime     datetime64[us]
lpep_dropoff_datetime    datetime64[us]
store_and_fwd_flag               object
RatecodeID                      float64
PULocationID                      int64
DOLocationID                      int64
passenger_count                 float64
trip_distance                   float64
fare_amount                     float64
extra                           float64
mta_tax                         float64
tip_amount                      float64
tolls_amount                    float64
ehail_fee                        object
improvement_surcharge           float64
total_amount                    float64
payment_type                    float64
trip_type                       float64
congestion_surcharge            float64
dtype: object

We see there are no text fields, they are all numbers. But some of those are categories or ID values:
- VendorID
- RatecodeID
- PULocationID
- DOLocationID
- store_and_fwd_flag 
- Trip_type
- payment_type

Numeric values:
- passenger_count
- trip_distance
- fare_amount
- extra
- mta_tax
- tip_amount
- tolls_amount
- improvement_surcharge
- total_amount
- congestion_surcharge

### Data Information

Duplicates

In [18]:
green22.duplicated().sum()

0

There are no duplicated rows in this dataset.

Missing and Null Values

In [19]:
green22.isna().sum()

VendorID                     0
lpep_pickup_datetime         0
lpep_dropoff_datetime        0
store_and_fwd_flag        9282
RatecodeID                9282
PULocationID                 0
DOLocationID                 0
passenger_count           9282
trip_distance                0
fare_amount                  0
extra                        0
mta_tax                      0
tip_amount                   0
tolls_amount                 0
ehail_fee                73718
improvement_surcharge        0
total_amount                 0
payment_type              9282
trip_type                 9283
congestion_surcharge      9282
dtype: int64

In all 73718 rows, the column 'ehail_fee' is completely empty. And that we have some missing values in the columns `store_and_fwd_flag`, 'RatecodeID', 'passenger_count', 'payment_type', 'trip_type' and 'congestion_surcharge'.

As the number of missing values is the same in these columns, it could be that they are in the same rows, we could see that in further analysis. For now, we could say that the number of null values is around the 12% of the dataset.

### Summary Statistics

In [20]:
numcolumns = ['passenger_count', 'trip_distance', 'fare_amount', 'extra', 'mta_tax', 'tip_amount', 'tolls_amount', 'improvement_surcharge', 'total_amount', 'congestion_surcharge'] 

green22[numcolumns].describe()

passenger_count  trip_distance   fare_amount         extra  \
count     64436.000000   73718.000000  73718.000000  73718.000000   
mean          1.283723     104.576467     15.671497      0.346503   
std           0.909935    3609.883882     14.004140      0.650320   
min           0.000000       0.000000    -85.000000     -4.500000   
25%           1.000000       1.170000      8.000000      0.000000   
50%           1.000000       2.100000     11.500000      0.000000   
75%           1.000000       3.910000     19.000000      0.500000   
max           8.000000  292061.660000    399.500000      4.500000   

            mta_tax    tip_amount  tolls_amount  improvement_surcharge  \
count  73718.000000  73718.000000  73718.000000           73718.000000   
mean       0.412261      2.115920      0.277911               0.297037   
std        0.193454      3.250391      1.412800               0.036460   
min       -0.500000     -0.960000      0.000000              -0.300000   
25%        0.500000      0.000000      0.000000               0.300000   
50%        0.500000      1.500000      0.000000               0.300000   
75%        0.500000      3.010000      0.000000               0.300000   
max        3.550000    222.250000     31.750000               0.300000   

       total_amount  congestion_surcharge  
count  73718.000000          64436.000000  
mean      19.793064              0.740933  
std       16.203584              1.220242  
min      -85.300000             -2.750000  
25%       10.300000              0.000000  
50%       15.360000              0.000000  
75%       23.727500              2.750000  
max      400.300000              2.750000

We can see there're some outliers, including negative values for amounts paid, and a maximum distance of 292061 miles.

### Data Distribution

Now, we'll analyze the distribution of pick up and drop off locations.

In [21]:
Newark_zone = [1]
JFK_zone = [132]
LaGuardia_zone = [138]
Manhattan_zones = [4, 12, 13, 24, 41, 42, 43, 45, 48, 50, 68, 74, 75, 79, 87, 88, 90, 103, 107, 113, 114, 116, 120, 125, 
             127, 128, 137, 140, 141, 142, 143, 144, 148, 151, 152, 153, 158, 161, 162, 163, 164, 166, 170, 186, 194,
               202, 209, 211, 224, 229, 230, 231, 232, 233, 234, 236, 237, 238, 239, 243, 244, 246, 249, 261, 262, 263]

In [22]:
manhattan_pu = green22.loc[green22.PULocationID.isin(Manhattan_zones)]
manhattan_pu

VendorID lpep_pickup_datetime lpep_dropoff_datetime store_and_fwd_flag  \
4             2  2022-06-01 00:35:38   2022-06-01 00:51:39                  N   
5             2  2022-06-01 00:28:58   2022-06-01 00:53:32                  N   
6             2  2022-06-01 00:17:24   2022-06-01 00:24:15                  N   
7             2  2022-06-01 00:57:08   2022-06-01 01:18:30                  N   
16            2  2022-06-01 00:37:41   2022-06-01 00:46:04                  N   
...         ...                  ...                   ...                ...   
73704         2  2022-06-30 21:05:00   2022-06-30 21:13:00               None   
73706         2  2022-06-30 21:33:00   2022-06-30 21:57:00               None   
73710         2  2022-06-30 21:09:00   2022-06-30 21:26:00               None   
73711         2  2022-06-30 21:06:00   2022-06-30 21:24:00               None   
73713         2  2022-06-30 22:33:00   2022-06-30 22:46:00               None   

       RatecodeID  PULocationID  DOLocationID  passenger_count  trip_distance  \
4             1.0            41           159              3.0           3.10   
5             1.0            42           216              1.0          16.18   
6             1.0            75           236              1.0           1.16   
7             1.0            75           235              1.0           6.00   
16            1.0           202             7              1.0           1.98   
...           ...           ...           ...              ...            ...   
73704         NaN            43           151              NaN           1.82   
73706         NaN            74            59              NaN           6.78   
73710         NaN            41           224              NaN           6.01   
73711         NaN            42           235              NaN           4.08   
73713         NaN            74           169              NaN           2.86   

       fare_amount  extra  mta_tax  tip_amount  tolls_amount ehail_fee  \
4            13.00    0.5      0.5        0.00          0.00      None   
5            43.50    0.5      0.5        0.00          6.55      None   
6             7.00    0.5      0.5        0.00          0.00      None   
7            20.00    0.5      0.5        5.00          0.00      None   
16            8.50    0.5      0.5       20.00          0.00      None   
...            ...    ...      ...         ...           ...       ...   
73704         9.90    0.0      0.0        2.39          0.00      None   
73706        16.93    0.0      0.0        0.00          0.00      None   
73710        22.87    0.0      0.0        4.37          0.00      None   
73711        16.70    0.0      0.0        0.00          0.00      None   
73713        18.66    0.0      0.0        2.30          0.00      None   

       improvement_surcharge  total_amount  payment_type  trip_type  \
4                        0.3         14.30           2.0        1.0   
5                        0.3         51.35           2.0        1.0   
6                        0.3         11.05           2.0        1.0   
7                        0.3         26.30           1.0        1.0   
16                       0.3         29.80           1.0        1.0   
...                      ...           ...           ...        ...   
73704                    0.3         12.59           NaN        NaN   
73706                    0.3         17.23           NaN        NaN   
73710                    0.3         30.29           NaN        NaN   
73711                    0.3         17.00           NaN        NaN   
73713                    0.3         21.26           NaN        NaN   

       congestion_surcharge  
4                      0.00  
5                      0.00  
6                      2.75  
7                      0.00  
16                     0.00  
...                     ...  
73704                   NaN  
73706                   NaN  
73710                   NaN  
73711              

In [23]:
manhattan_do = green22.loc[green22.DOLocationID.isin(Manhattan_zones)]
manhattan_do

VendorID lpep_pickup_datetime lpep_dropoff_datetime store_and_fwd_flag  \
6             2  2022-06-01 00:17:24   2022-06-01 00:24:15                  N   
18            2  2022-06-01 00:49:50   2022-06-01 00:55:32                  N   
20            2  2022-06-01 00:36:37   2022-06-01 00:42:29                  N   
22            2  2022-06-01 00:03:39   2022-06-01 00:14:59                  N   
29            2  2022-06-01 01:02:19   2022-06-01 01:14:39                  N   
...         ...                  ...                   ...                ...   
73702         2  2022-06-30 20:55:00   2022-06-30 21:07:00               None   
73703         2  2022-06-30 20:26:00   2022-06-30 20:40:00               None   
73704         2  2022-06-30 21:05:00   2022-06-30 21:13:00               None   
73710         2  2022-06-30 21:09:00   2022-06-30 21:26:00               None   
73715         2  2022-06-30 22:10:00   2022-06-30 22:55:00               None   

       RatecodeID  PULocationID  DOLocationID  passenger_count  trip_distance  \
6             1.0            75           236              1.0           1.16   
18            1.0           166           166              1.0           0.95   
20            1.0            75           238              1.0           1.19   
22            1.0            74           166              1.0           1.58   
29            1.0            41           140              1.0           3.52   
...           ...           ...           ...              ...            ...   
73702         NaN            41           142              NaN           2.89   
73703         NaN            41           239              NaN           2.64   
73704         NaN            43           151              NaN           1.82   
73710         NaN            41           224              NaN           6.01   
73715         NaN           106           137              NaN           8.47   

       fare_amount  extra  mta_tax  tip_amount  tolls_amount ehail_fee  \
6             7.00    0.5      0.5        0.00           0.0      None   
18            5.50    0.5      0.5        1.36           0.0      None   
20            6.50    0.5      0.5        2.00           0.0      None   
22            9.00    0.5      0.5        1.85           0.0      None   
29           13.00    0.5      0.5        0.00           0.0      None   
...            ...    ...      ...         ...           ...       ...   
73702        12.58    0.0      0.0        3.53           0.0      None   
73703        13.17    0.0      0.0        0.00           0.0      None   
73704         9.90    0.0      0.0        2.39           0.0      None   
73710        22.87    0.0      0.0        4.37           0.0      None   
73715        32.69    0.0      0.0        4.97           0.0      None   

       improvement_surcharge  total_amount  payment_type  trip_type  \
6                        0.3         11.05           2.0        1.0   
18                       0.3          8.16           1.0        1.0   
20                       0.3         12.55           1.0        1.0   
22                       0.3         12.15           1.0        1.0   
29                       0.3         17.05           2.0        1.0   
...                      ...           ...           ...        ...   
73702                    0.3         19.16           NaN        NaN   
73703                    0.3         16.22           NaN        NaN   
73704                    0.3         12.59           NaN        NaN   
73710                    0.3         30.29           NaN        NaN   
73715                    0.3         40.71           NaN        NaN   

       congestion_surcharge  
6                      2.75  
18                     0.00  
20                     2.75  
22                     0.00  
29                     2.75  
...                     ...  
73702                   NaN  
73703                   NaN  
73704                   NaN  
73710              

In this dataset, around the 55% of the trips have a Manhattan zone in their pick up or drop off location.

In [24]:
to_newark = green22['trip_distance'].loc[green22.PULocationID.isin(Manhattan_zones) & green22.DOLocationID.isin(Newark_zone)]
to_newark.describe()

count    22.000000
mean     26.088182
std       4.549634
min      14.270000
25%      23.890000
50%      26.455000
75%      27.455000
max      35.520000
Name: trip_distance, dtype: float64

In [25]:
to_jfk = green22['trip_distance'].loc[green22.PULocationID.isin(Manhattan_zones) & green22.DOLocationID.isin(JFK_zone)]
to_jfk.describe()

count    404.000000
mean      18.053267
std        2.951431
min        0.000000
25%       17.607500
50%       18.300000
75%       19.060000
max       23.070000
Name: trip_distance, dtype: float64

In [26]:
to_laguardia = green22['trip_distance'].loc[green22.PULocationID.isin(Manhattan_zones) & green22.DOLocationID.isin(LaGuardia_zone)]
to_laguardia.describe()

count       619.000000
mean        255.752100
std        6164.966567
min           0.000000
25%           6.835000
50%           7.830000
75%           8.600000
max      153390.540000
Name: trip_distance, dtype: float64

### Green_tripdata_2023-06

Importing the dataset as a DataFrame

In [27]:
green23= pd.read_parquet('Datasets\\green_tripdata_2023-06.parquet')

### Data Dictionary:
- The green_tripdata_2023-06 and green_tripdata_2023-06 share the same and dictionary.

In [28]:
green23.head()

VendorID lpep_pickup_datetime lpep_dropoff_datetime store_and_fwd_flag  \
0         2  2023-06-01 00:32:25   2023-06-01 00:39:06                  N   
1         2  2023-06-01 00:39:07   2023-06-01 00:51:59                  N   
2         2  2023-06-01 00:35:59   2023-06-01 00:57:06                  N   
3         2  2023-06-01 00:50:29   2023-06-01 01:00:32                  N   
4         2  2023-06-01 00:15:15   2023-06-01 00:52:40                  N   

   RatecodeID  PULocationID  DOLocationID  passenger_count  trip_distance  \
0         1.0            74            42              1.0           0.84   
1         1.0            41           229              1.0           4.05   
2         1.0            97            89              1.0           4.26   
3         1.0            75           263              1.0           1.79   
4         1.0            33            48              1.0           6.40   

   fare_amount  extra  mta_tax  tip_amount  tolls_amount  ehail_fee  \
0          7.9    1.0      0.5        2.08           0.0        NaN   
1         19.1    1.0      0.5        3.65           0.0        NaN   
2         23.3    1.0      0.5        5.16           0.0        NaN   
3         12.1    1.0      0.5        3.47           0.0        NaN   
4         37.3    1.0      0.5        0.00           0.0        NaN   

   improvement_surcharge  total_amount  payment_type  trip_type  \
0                    1.0         12.48           1.0        1.0   
1                    1.0         28.00           1.0        1.0   
2                    1.0         30.96           1.0        1.0   
3                    1.0         20.82           1.0        1.0   
4                    1.0         42.55           1.0        1.0   

   congestion_surcharge  
0                  0.00  
1                  2.75  
2                  0.00  
3                  2.75  
4                  2.75

We see there are the same columns than the 2022 dataset, so the dictionary works for both.

Also, the data types are equal.

### Shape and Dimension

In [29]:
green23.shape

(65550, 20)

Column Names and Data Types

In [30]:
green23.dtypes

VendorID                          int32
lpep_pickup_datetime     datetime64[us]
lpep_dropoff_datetime    datetime64[us]
store_and_fwd_flag               object
RatecodeID                      float64
PULocationID                      int32
DOLocationID                      int32
passenger_count                 float64
trip_distance                   float64
fare_amount                     float64
extra                           float64
mta_tax                         float64
tip_amount                      float64
tolls_amount                    float64
ehail_fee                       float64
improvement_surcharge           float64
total_amount                    float64
payment_type                    float64
trip_type                       float64
congestion_surcharge            float64
dtype: object

In [31]:
green23.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 65550 entries, 0 to 65549
Data columns (total 20 columns):
 #   Column                 Non-Null Count  Dtype         
---  ------                 --------------  -----         
 0   VendorID               65550 non-null  int32         
 1   lpep_pickup_datetime   65550 non-null  datetime64[us]
 2   lpep_dropoff_datetime  65550 non-null  datetime64[us]
 3   store_and_fwd_flag     60359 non-null  object        
 4   RatecodeID             60359 non-null  float64       
 5   PULocationID           65550 non-null  int32         
 6   DOLocationID           65550 non-null  int32         
 7   passenger_count        60359 non-null  float64       
 8   trip_distance          65550 non-null  float64       
 9   fare_amount            65550 non-null  float64       
 10  extra                  65550 non-null  float64       
 11  mta_tax                65550 non-null  float64       
 12  tip_amount             65550 non-null  float64       
 13  t

### Data Information

Duplicates

In [32]:
green23.duplicated().sum()

0

There are no duplicated rows in this dataset.

Missing and Null Values

In [33]:
green23.isna().sum()

VendorID                     0
lpep_pickup_datetime         0
lpep_dropoff_datetime        0
store_and_fwd_flag        5191
RatecodeID                5191
PULocationID                 0
DOLocationID                 0
passenger_count           5191
trip_distance                0
fare_amount                  0
extra                        0
mta_tax                      0
tip_amount                   0
tolls_amount                 0
ehail_fee                65550
improvement_surcharge        0
total_amount                 0
payment_type              5191
trip_type                 5191
congestion_surcharge      5191
dtype: int64

In all 65550 rows, the column 'ehail_fee' is completely empty. And that we have some missing values in the columns 'store_and_fwd_flag', 'RatecodeID', 'passenger_count', 'payment_type', 'trip_type' and 'congestion_surcharge'.


As the number of missing values is the same in these columns, it could be that they are in the same rows, we could see that in further analysis. For now, we could say that the number of null values is around the 8% of the dataset.

### Summary Satistics and Data Distribution

We'll analyze the distribution of the columns with numeric values. Excluding the ones that are categorical, IDs or dates.

In [34]:
numcolumns = ['passenger_count', 'trip_distance', 'fare_amount', 'extra', 'mta_tax', 'tip_amount', 'tolls_amount', 'improvement_surcharge', 'total_amount', 'congestion_surcharge'] 

green23[numcolumns].describe()

passenger_count  trip_distance   fare_amount         extra  \
count     60359.000000   65550.000000  65550.000000  65550.000000   
mean          1.283007      22.430450     18.892448      0.882613   
std           0.904936    1114.401748     22.605564      1.327904   
min           0.000000       0.000000   -240.000000     -2.500000   
25%           1.000000       1.180000     10.000000      0.000000   
50%           1.000000       1.980000     14.200000      0.000000   
75%           1.000000       3.600000     21.980000      1.000000   
max           9.000000  179375.160000   4003.000000     10.000000   

            mta_tax    tip_amount  tolls_amount  improvement_surcharge  \
count  65550.000000  65550.000000  65550.000000           65550.000000   
mean       0.561037      2.415067      0.253476               0.985623   
std        0.381443      3.308543      1.355565               0.132579   
min       -0.500000     -0.900000      0.000000              -1.000000   
25%        0.500000      0.000000      0.000000               1.000000   
50%        0.500000      1.940000      0.000000               1.000000   
75%        0.500000      3.690000      0.000000               1.000000   
max        1.500000     96.300000     29.500000               1.000000   

       total_amount  congestion_surcharge  
count  65550.000000          60359.000000  
mean      24.530728              0.746947  
std       24.247960              1.223115  
min     -241.000000             -2.750000  
25%       13.640000              0.000000  
50%       19.500000              0.000000  
75%       29.000000              2.750000  
max     4004.500000              2.750000

We can see there're some outliers, including negative values for amounts paid, and a maximum distance of 179375 miles. Also, it's noticeable that the standard deviation of the total amount is really big, considering the value of the 75% qt.